neural_modelling/src/neuron/plasticity/stdp/synapse_dynamics_stdp_izhikevich_neuromodulation.c

```diff
@@ -0,0 +1,483 @@
+
+    // Load timing dependence data
+    address_t weight_region_address = timing_initialise(address);
+    if (address == NULL) {
+        return false;
+    }
+
+    // Read Izhikevich weight update equation constant component
+    weight_update_constant_component = *weight_region_address++;
+
+    // Load weight dependence data
+    address_t weight_result = weight_initialise(
+        weight_region_address, ring_buffer_to_input_buffer_left_shifts);
+    if (weight_result == NULL) {
+        return false;
+    }
+
+    post_event_history = post_events_init_buffers(n_neurons);
+    if (post_event_history == NULL) {
+        return false;
+    }
+
+    return true;
+}
+
+//---------------------------------------
+// Synaptic row plastic-region implementation
+//---------------------------------------
+static inline plastic_synapse_t* plastic_synapses(address_t plastic) {
+    const uint32_t pre_event_history_size_words =
+        sizeof(pre_event_history_t) / sizeof(uint32_t);
+    static_assert(pre_event_history_size_words * sizeof(uint32_t) == sizeof(pre_event_history_t),
+        "Size of pre_event_history_t structure should be a multiple of 32-bit words");
+
+    return (plastic_synapse_t*)(&plastic[pre_event_history_size_words]);
+}
+
+//---------------------------------------
+static inline pre_event_history_t *plastic_event_history(address_t plastic) {
+    return (pre_event_history_t*)(&plastic[0]);
+}
+
+//---------------------------------------
+static inline index_t sparse_axonal_delay(uint32_t x) {
+    return ((x >> SYNAPSE_DELAY_TYPE_INDEX_BITS) & SYNAPSE_AXONAL_DELAY_MASK);
+}
+
+//---------------------------------------
+void synapse_dynamics_process_post_synaptic_event(
+        uint32_t time, index_t neuron_index) {
+    log_debug("Adding post-synaptic event to trace at time:%u", time);
+
+    // Add post-event
+    post_event_history_t *history = &post_event_history[neuron_index];
+    const uint32_t last_post_time = history->times[history->count_minus_one];
+    const post_trace_t last_post_trace =
+        history->traces[history->count_minus_one];
+    post_events_add(time, history, timing_add_post_spike(time, last_post_time,
+                                                     last_post_trace), false);
+}
+
+//--------------------------------------
+void synapse_dynamics_process_neuromodulator_event(
+        uint32_t time, int32_t concentration, uint32_t neuron_index,
+        uint32_t synapse_type) {
+    log_debug(
+        "Adding neuromodulation event to trace at time:%u concentration:%d",
+        time, concentration);
+
+    // Get post event history of this neuron
+    post_event_history_t *history = &post_event_history[neuron_index];
+    const uint32_t last_post_time = history->times[history->count_minus_one];
+    const post_trace_t last_post_trace =
+        history->traces[history->count_minus_one];
+
+    // Update neuromodulator level reaching this post synaptic neuron
```

Comment:
@oliverrhodes note that this comment is misleading - we need to know not just dopamine level at the post-synaptic neuron, but a history of dopamine spikes in order to be able to integrate them (see post_events_with_da.h). I will update the comments.